readthedocs/projects/tasks/builds.py

```diff
@@ -513,55 +512,84 @@ def on_failure(self, exc, task_id, args, kwargs, einfo):
         # Update build object
         self.data.build['success'] = False
 
-    def on_success(self, retval, task_id, args, kwargs):
-        time_before_store_build_artifacts = timezone.now()
-        # Store build artifacts to storage (local or cloud storage)
-        #
-        # TODO: move ``store_build_artifacts`` inside ``execute`` so we can
-        # handle exceptions properly on ``on_failure``
-        self.store_build_artifacts()
-        log.info(
-            "Store build artifacts finished.",
-            time=(timezone.now() - time_before_store_build_artifacts).seconds,
-        )
+    def get_valid_artifact_types(self):
+        """
+        Return a list of all the valid artifact types for this build.
 
-        # HTML are built successfully.
-        html = os.path.exists(
-            self.data.project.artifact_path(
-                version=self.data.version.slug, type_="html"
-            )
-        )
-        localmedia = os.path.exists(
-            self.data.project.artifact_path(
-                version=self.data.version.slug, type_="htmlzip"
-            )
-        )
-        pdf = os.path.exists(
-            self.data.project.artifact_path(version=self.data.version.slug, type_="pdf")
-        )
-        epub = os.path.exists(
-            self.data.project.artifact_path(
-                version=self.data.version.slug, type_="epub"
+        It performs the following checks on each output format type path:
+         - it exists
+         - it is a directory
+         - does not contains more than 1 files (only PDF, HTMLZip, ePUB)
+
+        TODO: remove the limitation of only 1 file.
+        Add support for multiple PDF files in the output directory and
+        grab them by using glob syntaxt between other files that could be garbage.
+        """
+        valid_artifacts = []
+        for artifact_type in ARTIFACT_TYPES:
+            artifact_directory = self.data.project.artifact_path(
+                version=self.data.version.slug,
+                type_=artifact_type,
             )
-        )
+            if not os.path.exists(artifact_directory):
+                # There is no output directory.
+                # Skip this format.
+                continue
+
+            if not os.path.isdir(artifact_directory):
+                log.error(
+                    "The output path is not a directory.",
+                    output_format=artifact_type,
+                )
+                raise BuildUserError(
+                    BuildUserError.BUILD_OUTPUT_IS_NOT_A_DIRECTORY.format(
+                        artifact_type=artifact_type
+                    )
+                )
+
+            # Check if there are multiple files on artifact directories.
+            # These output format does not support multiple files yet.
+            # In case multiple files are found, the upload for this format is not performed.
+            if artifact_type in ("htmlzip", "epub", "pdf"):
```

Comment:
Address Eric's feedback for this https://github.com/readthedocs/readthedocs.org/pull/9940#discussion_r1087017048